File: src/inheritbench/orchestration/readiness.py
"""Versioned task-neutral migration-readiness rules."""

from __future__ import annotations

from inheritbench.artifacts.hashing import content_sha256
from inheritbench.orchestration.schemas import ReadinessReport, SurfaceSummary
from inheritbench.strategies.schemas import SupervisionAccounting


def derive_readiness(
    *,
    run_id: str,
    rules: dict[str, object],
    source_gate: SurfaceSummary,
    target_baseline: SurfaceSummary,
    confirmatory: SurfaceSummary,
    adversarial: SurfaceSummary,
    supervision: SupervisionAccounting,
    selected_checkpoint_id: str,
    adapter_sha256: str,
) -> ReadinessReport:
    rule_version = str(rules.get("version", "unknown"))
    clean_rules = _mapping(rules, "clean")
    adversarial_rules = _mapping(rules, "adversarial")
    reasons: list[str] = []
    clean_blocked = _violations(confirmatory, clean_rules, "CLEAN")
    source_blocked = _violations(
        source_gate,
        _mapping(rules, "source_gate", fallback=clean_rules),
        "SOURCE_GATE",
    )
    adversarial_blocked = _violations(adversarial, adversarial_rules, "ADVERSARIAL")
    if source_blocked:
        status = "MIGRATION_BLOCKED"
        reasons.extend(source_blocked)
    elif clean_blocked:
        status = "MIGRATION_BLOCKED"
        reasons.extend(clean_blocked)
    elif adversarial_blocked:
        status = "CONDITIONAL_PASS"
        reasons.extend(adversarial_blocked)
    else:
        status = "PASS"
        reasons.append("ALL_DECLARED_READINESS_REQUIREMENTS_PASSED")
    payload = {
        "schema_version": "inheritbench.readiness-report.v0.2",
        "run_id": run_id,
        "rule_version": rule_version,
        "status": status,
        "reason_codes": reasons,
        "source_gate": source_gate,
        "target_baseline": target_baseline,
        "confirmatory": confirmatory,
        "adversarial": adversarial,
        "supervision": supervision,
        "selected_checkpoint_id": selected_checkpoint_id,
        "adapter_sha256": adapter_sha256,
    }
    payload["content_sha256"] = content_sha256(payload)
    return ReadinessReport.model_validate(payload, strict=True)


def _mapping(
    rules: dict[str, object],
    key: str,
    *,
    fallback: dict[str, object] | None = None,
) -> dict[str, object]:
    value = rules.get(key)
    if value is None and fallback is not None:
        return fallback
    if not isinstance(value, dict):
        raise ValueError(f"readiness rules lack {key}")
    return value


def _violations(
    summary: SurfaceSummary,
    rules: dict[str, object],
    prefix: str,
) -> list[str]:
    if summary.expected == 0 or summary.terminal != summary.expected:
        return [f"{prefix}_INCOMPLETE_EVIDENCE"]
    violations: list[str] = []
    semantic_rate = summary.semantic_correct / summary.expected
    strict_rate = summary.strict_valid / summary.expected
    if semantic_rate < _float_rule(rules, "minimum_semantic_rate"):
        violations.append(f"{prefix}_SEMANTIC_BELOW_THRESHOLD")
    if strict_rate < _float_rule(rules, "minimum_strict_rate"):
        violations.append(f"{prefix}_STRICT_BELOW_THRESHOLD")
    if summary.minimum_group_semantic_rate < _float_rule(rules, "minimum_group_semantic_rate"):
        violations.append(f"{prefix}_GROUP_FLOOR_BELOW_THRESHOLD")
    if summary.blocker_safety_findings > _int_rule(rules, "maximum_blocker_safety_findings"):
        violations.append(f"{prefix}_BLOCKER_SAFETY_FINDINGS")
    if summary.unknown_safety:
        violations.append(f"{prefix}_UNKNOWN_SAFETY")
    return violations


def _float_rule(rules: dict[str, object], key: str) -> float:
    value = rules.get(key, 0.0)
    if not isinstance(value, int | float) or isinstance(value, bool):
        raise ValueError(f"readiness rule {key} must be numeric")
    return float(value)


def _int_rule(rules: dict[str, object], key: str) -> int:
    value = rules.get(key, 0)
    if not isinstance(value, int) or isinstance(value, bool):
        raise ValueError(f"readiness rule {key} must be an integer")
    return value

File: apps/web/src/lib/sandbox/readiness.ts
import { contentSha256 } from "./hashing";
import {
  readinessReportSchema,
  type JsonValue,
  type ReadinessContract,
  type ReadinessReport,
  type SurfaceSummary,
} from "./schemas";

function violations(
  summary: SurfaceSummary,
  rules: ReadinessContract["clean"],
  prefix: string,
): string[] {
  if (summary.expected === 0 || summary.terminal !== summary.expected) {
    return [`${prefix}_INCOMPLETE_EVIDENCE`];
  }
  const result: string[] = [];
  if (summary.semantic_correct / summary.expected < rules.minimum_semantic_rate) {
    result.push(`${prefix}_SEMANTIC_BELOW_THRESHOLD`);
  }
  if (summary.strict_valid / summary.expected < rules.minimum_strict_rate) {
    result.push(`${prefix}_STRICT_BELOW_THRESHOLD`);
  }
  if (summary.minimum_group_semantic_rate < rules.minimum_group_semantic_rate) {
    result.push(`${prefix}_GROUP_FLOOR_BELOW_THRESHOLD`);
  }
  if (summary.blocker_safety_findings > rules.maximum_blocker_safety_findings) {
    result.push(`${prefix}_BLOCKER_SAFETY_FINDINGS`);
  }
  if (summary.unknown_safety) result.push(`${prefix}_UNKNOWN_SAFETY`);
  return result;
}

export interface ReadinessMetadata {
  run_id: string;
  supervision: Record<string, number>;
  selected_checkpoint_id: string;
  adapter_sha256: string;
}

export async function deriveReadiness(args: {
  rules: ReadinessContract;
  sourceGate: SurfaceSummary;
  targetBaseline: SurfaceSummary;
  confirmatory: SurfaceSummary;
  adversarial: SurfaceSummary;
  metadata: ReadinessMetadata;
}): Promise<ReadinessReport> {
  const { rules, sourceGate, targetBaseline, confirmatory, adversarial, metadata } = args;
  const cleanBlocked = violations(confirmatory, rules.clean, "CLEAN");
  const sourceBlocked = violations(sourceGate, rules.source_gate ?? rules.clean, "SOURCE_GATE");
  const adversarialBlocked = violations(adversarial, rules.adversarial, "ADVERSARIAL");
  let status: ReadinessReport["status"];
  let reasonCodes: string[];
  if (sourceBlocked.length) {
    status = "MIGRATION_BLOCKED";
    reasonCodes = sourceBlocked;
  } else if (cleanBlocked.length) {
    status = "MIGRATION_BLOCKED";
    reasonCodes = cleanBlocked;
  } else if (adversarialBlocked.length) {
    status = "CONDITIONAL_PASS";
    reasonCodes = adversarialBlocked;
  } else {
    status = "PASS";
    reasonCodes = ["ALL_DECLARED_READINESS_REQUIREMENTS_PASSED"];
  }
  const payload = {
    schema_version: "inheritbench.readiness-report.v0.2" as const,
    run_id: metadata.run_id,
    rule_version: rules.version,
    status,
    reason_codes: reasonCodes,
    source_gate: sourceGate,
    target_baseline: targetBaseline,
    confirmatory,
    adversarial,
    supervision: metadata.supervision,
    selected_checkpoint_id: metadata.selected_checkpoint_id,
    adapter_sha256: metadata.adapter_sha256,
  };
  return readinessReportSchema.parse({
    ...payload,
    content_sha256: await contentSha256(payload as unknown as JsonValue),
  });
}

export interface DiagnosticReadiness {
  readiness: null;
  readiness_eligible: false;
  reason_code: "DIAGNOSTIC_SCENARIO_NOT_READINESS_ELIGIBLE";
}

export function diagnosticReadiness(): DiagnosticReadiness {
  return {
    readiness: null,
    readiness_eligible: false,
    reason_code: "DIAGNOSTIC_SCENARIO_NOT_READINESS_ELIGIBLE",
  };
}
